File: src/app/types/bill.ts
// Bill types — Bill, BillMetadata, ParsingStatus

export type ParsingStatus = "pending" | "processing" | "completed" | "failed";
export type DocumentType = "provider_bill" | "hospital_statement" | "eob" | "denial_letter" | "itemized_statement" | "unknown";
export type VisitType = "emergency" | "outpatient" | "inpatient" | "imaging" | "other";

export interface BillDocument {
  s3_key: string;
  filename: string;
  content_type: string;
  uploaded_at: string;
}

export interface ConfidenceScores {
  overall: number;
  fields: Record<string, number>;
}

export interface Bill {
  _id: string;
  user_id: string;
  provider: string | null;
  facility: string | null;
  visit_type: VisitType;
  service_date_range: { start: string | null; end: string | null };
  total_billed: number | null;
  total_allowed: number | null;
  total_insurance_paid: number | null;
  patient_balance: number | null;
  insurance_provider: string | null;
  document_type: DocumentType;
  documents: BillDocument[];
  parsing_status: ParsingStatus;
  confidence_scores: ConfidenceScores;
  plain_language_summary: string | null;
  user_notes: string | null;
  created_at: string;
  updated_at: string;
}

File: backend/models/bill.py
# Bill Pydantic Models

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class ParsingStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class DocumentType(str, Enum):
    PROVIDER_BILL = "provider_bill"
    HOSPITAL_STATEMENT = "hospital_statement"
    EOB = "eob"
    DENIAL_LETTER = "denial_letter"
    ITEMIZED_STATEMENT = "itemized_statement"
    UNKNOWN = "unknown"


class VisitType(str, Enum):
    EMERGENCY = "emergency"
    OUTPATIENT = "outpatient"
    INPATIENT = "inpatient"
    IMAGING = "imaging"
    OTHER = "other"


class DateRange(BaseModel):
    start: Optional[str] = None
    end: Optional[str] = None


class BillDocument(BaseModel):
    s3_key: str
    filename: str
    content_type: str
    uploaded_at: datetime


class ConfidenceScores(BaseModel):
    overall: float = 0.0
    fields: dict = Field(default_factory=dict)


class BillResponse(BaseModel):
    id: str = Field(alias="_id")
    user_id: str
    provider: Optional[str] = None
    facility: Optional[str] = None
    visit_type: Optional[VisitType] = None
    service_date_range: DateRange = Field(default_factory=DateRange)
    total_billed: Optional[float] = None
    total_allowed: Optional[float] = None
    total_insurance_paid: Optional[float] = None
    patient_balance: Optional[float] = None
    insurance_provider: Optional[str] = None
    document_type: Optional[DocumentType] = None
    documents: List[BillDocument] = Field(default_factory=list)
    parsing_status: ParsingStatus = ParsingStatus.PENDING
    confidence_scores: ConfidenceScores = Field(default_factory=ConfidenceScores)
    plain_language_summary: Optional[str] = None
    user_notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        populate_by_name = True
